OTA-1427: Recompute node insights when MC/MCP changes

hongkailiu opened the pull request:
This is to address the following comments from the review [1]:

- [ ] Decide what to do with the code we copied from MCO (https://github.com/openshift/cluster-version-operator/pull/1136#discussion_r1926022555)
- [x] Avoid iterating all MCP resource and re-computing selectors on every Node-triggered sync (https://github.com/openshift/cluster-version-operator/pull/1136#discussion_r1927432583, https://github.com/openshift/cluster-version-operator/pull/1136#discussion_r1932496850)
- [x] Recompute Node insights when MCP selectors change (changed selectors can change MCP membership, potentially making Node insights stale) (https://github.com/openshift/cluster-version-operator/pull/1136#discussion_r1932496850)
- [x] Avoid iterating all MC resources on every Node-triggered sync (https://github.com/openshift/cluster-version-operator/pull/1136#discussion_r1932496850)
- [x] Decide how to gracefully handle Nodes that are not matched by any MCP selector (https://github.com/openshift/cluster-version-operator/pull/1136#discussion_r1928890801, https://github.com/openshift/cluster-version-operator/pull/1136#discussion_r1932517120)
- [ ] Polishing condition reasons, messages and the "summary" insight message (https://github.com/openshift/cluster-version-operator/pull/1136#discussion_r1932555649, https://github.com/openshift/cluster-version-operator/pull/1136#discussion_r1932569171, https://github.com/openshift/cluster-version-operator/pull/1136#discussion_r1932571522)

[1]. https://github.com/openshift/cluster-version-operator/pull/1136#issuecomment-2621934987

hongkailiu commented:
/test e2e-agnostic-usc-devpreview

hongkailiu commented:
/test e2e-agnostic-usc-devpreview

hongkailiu commented:
The init seems not working ... 🤔 

https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/openshift_cluster-version-operator/1144/pull-ci-openshift-cluster-version-operator-main-e2e-agnostic-usc-devpreview/1885848843463954432/artifacts/e2e-agnostic-usc-devpreview/gather-extra/artifacts/pods/openshift-update-status-controller_update-status-controller-65c495d5c6-4ljd7_update-status-controller.log

```
I0202 01:52:01.950445       1 nodeinformer.go:278] Stored 0 machineConfigPools in the cache
I0202 01:52:01.950524       1 nodeinformer.go:290] Stored 0 machineConfig versions in the cache
```

hongkailiu commented:
```console
$ curl -s https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/openshift_cluster-version-operator/1144/pull-ci-openshift-cluster-version-operator-main-e2e-agnostic-usc-devpreview/1885905947818201088/artifacts/e2e-agnostic-usc-devpreview/gather-extra/artifacts/pods/openshift-update-status-controller_update-status-controller-844f579bfd-mxg7z_update-status-controller.log | rg 'Syncing with key|which|Cached|Stored|Reconciling all nodes' | head -n 10
I0202 06:06:53.968786       1 nodeinformer.go:282] Stored 0 machineConfigPools in the cache
I0202 06:06:53.968954       1 nodeinformer.go:294] Stored 0 machineConfig versions in the cache
I0202 06:06:54.183789       1 nodeinformer.go:107] NI :: Syncing with key MachineConfigPool/master
I0202 06:06:54.183939       1 nodeinformer.go:244] Cached MachineConfigPool master with selector node-role.kubernetes.io/master=
I0202 06:06:54.184005       1 nodeinformer.go:182] Reconciling all nodes as machine config pool "master" is refreshed
I0202 06:06:54.184104       1 nodeinformer.go:107] NI :: Syncing with key MachineConfigPool/worker
I0202 06:06:54.184171       1 nodeinformer.go:244] Cached MachineConfigPool worker with selector node-role.kubernetes.io/worker=
I0202 06:06:54.184224       1 nodeinformer.go:182] Reconciling all nodes as machine config pool "worker" is refreshed
I0202 06:06:54.184279       1 nodeinformer.go:107] NI :: Syncing with key Node/ci-op-n2q0th1r-8c750-5r658-master-0
I0202 06:06:54.200901       1 nodeinformer.go:107] NI :: Syncing with key Node/ci-op-n2q0th1r-8c750-5r658-master-1
```

In the init phase, the lister was not synced yet. However, it caught up by the events on MCPs.
It is safe to to remove `initializeCaches()` as it is useless.

https://github.com/openshift/cluster-version-operator/pull/1144/files#diff-c5a41cb83200e13544e0509a5e77c180e16a1d70d7411a79cac2bc107e89950fR271-R298

hongkailiu commented on pkg/updatestatus/nodeinformer.go:
```diff
@@ -94,26 +119,16 @@ func (c *nodeInformerController) sync(ctx context.Context, syncCtx factory.SyncC
+			// At any time, any node belongs to one and only one MCP.
+			// Otherwise, it indicates a configuration error on the cluster.
+			klog.Errorf("Failed to determine which machine config pool the node %s belongs to in %d pools from the cache",
+				node.Name, len(c.machineConfigPoolCache))
```
I am bouncing between showing this configuration error in some status API that is exposed to cluster admins or only logging it (what the code does at the moment).

On the one hand, USC is not responsible for displaying those configuration errors to admins. On the other, this error might impact cluster upgrade. I do not know how MCO react on this error as MCO has the assumption that every node belongs to ONE and only MCP as all time.

Maybe we should have a special kind of reason for those nodes, say "orphaned" (similar to "paused")?

```go
{
  Type:               "Updating",
  Status:             "False",
  Reason:            "Orphaned",
  Message:         "The node <node> is orphaned as it does not belong to any machine config pool which is a configuration error and need to be fixed."
  LastTransitionTime: now,
},
```

hongkailiu replied:
Since it is likely to get more involved to handle the case, I will isolate it to another PR.
I have already remove the item from the PR description.

hongkailiu replied:
Thinking aloud:

"the node is not managed by the MCO" was not a thing in oc-cli for the status cmd. Do we want to consider it now?

DavidHurta replied:
From [OpenShift documentation](https://docs.openshift.com/container-platform/4.17/machine_configuration/index.html#architecture-machine-config-pools_machine-config-overview):

>... 
> If you remove the worker label from a node and apply the infra label without grouping it in a custom pool, the node is not recognized by the MCO and is unmanaged by the cluster.
> ...

So it seems like a node, which is not managed by the MCO, is not part of a cluster upgrade as it is unmanaged. It seems like such a node is not inherently misconfigured; however, it may be potentially prone to some issues as the node is managed by a cluster administrator (e.g. kubelet version skew).

---

> "the node is not managed by the MCO" was not a thing in oc-cli for the status cmd. Do we want to consider it now?

Good question. I feel like it depends on whether such nodes might impact a cluster upgrade.

I can imagine a `kube-apiserver` being upgraded, kubelet of an unmanaged node becomes unhappy due to a version skew, the status of its node `Ready` condition becomes `False` -> [`the node is not healthy and is not accepting pods`](https://kubernetes.io/docs/reference/node/node-status/#condition), cluster upgrade successfully completes; however, it has resulted in degradation of a node.

The responsibility of the upgrade status feature is on upgrades and not on the state of a cluster.
I think it depends on whether an unmanaged node might impact a cluster upgrade. Which I am not sure. If there are any issues with such nodes, impacted components should ideally report such issues clearly. I think it is okay for the moment to ignore such nodes (as we do in the CLI) as they are not part of a cluster upgrade. If the nodes are to impact cluster upgrade and impacted components are not able to report such an issue, we can always revisit the topic. However, it would be nice to ignore such nodes explicitly and not by accident :sweat_smile:

petr-muller replied:
Right now `oc adm upgrade status` only shows nodes with the assumption that they belong to a MCP, if a Node does not, we do not show it. Lets not make our lives harder by adding new functionality (yet). If we hear Nodes outside MCP are a real problem encountered during updates, we can add code that handles them in an way appropriate to the problems they cause. For now I'd just ignore nodes that do not belong to a MCP, just like CLI does.

hongkailiu replied:
The 4th commit is to handle the error case of no matching MCP (also revised the PR description to include it):

- Add comments so we wont forget why we are doing it this way
- Add a log for debugging
- Ignore the insight for these nodes

hongkailiu commented:
/retest-required

hongkailiu commented:
```
curl -s https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/openshift_cluster-version-operator/1144/pull-ci-openshift-cluster-version-operator-main-e2e-agnostic-usc-devpreview/1886150916344647680/artifacts/e2e-agnostic-usc-devpreview/gather-extra/artifacts/pods/openshift-update-status-controller_update-status-controller-56468d56d6-trf86_update-status-controller.log | rg 'Syncing with key|which|Cached|Stored|Reconciling all nodes' | head -n 10
I0202 22:07:59.245643       1 nodeinformer.go:103] NI :: Syncing with key MachineConfigPool/master
I0202 22:07:59.245746       1 nodeinformer.go:243] Cached MachineConfigPool master with selector node-role.kubernetes.io/master=
I0202 22:07:59.245776       1 nodeinformer.go:178] Reconciling all nodes as machine config pool "master" is refreshed
I0202 22:07:59.245842       1 nodeinformer.go:103] NI :: Syncing with key MachineConfigPool/worker
I0202 22:07:59.245925       1 nodeinformer.go:243] Cached MachineConfigPool worker with selector node-role.kubernetes.io/worker=
I0202 22:07:59.245963       1 nodeinformer.go:178] Reconciling all nodes as machine config pool "worker" is refreshed
I0202 22:07:59.246009       1 nodeinformer.go:103] NI :: Syncing with key Node/ci-op-89pfjfg5-8c750-p4r2f-master-0
I0202 22:07:59.254500       1 nodeinformer.go:103] NI :: Syncing with key Node/ci-op-89pfjfg5-8c750-p4r2f-master-1
I0202 22:07:59.268333       1 nodeinformer.go:103] NI :: Syncing with key Node/ci-op-89pfjfg5-8c750-p4r2f-master-2
I0202 22:07:59.280301       1 nodeinformer.go:103] NI :: Syncing with key Node/ci-op-89pfjfg5-8c750-p4r2f-worker-westus-rx8ls
```

The log looks good.

hongkailiu commented:
/test e2e-hypershift

hongkailiu commented:
/cc @petr-muller

DavidHurta commented on pkg/updatestatus/nodeinformer.go:
```diff
@@ -393,8 +511,16 @@ func nodeInformerControllerQueueKeys(object runtime.Object) []string {
 	}
 }
+
+func nodeNameToQueueKey(nodeName string) string {
+	return fmt.Sprintf("%s/%s", nodeKindName, nodeName)
+}
```
Or potentially add `machineConfigNameToQueueKey` and `machineConfigPoolNameToQueueKey` functions?

DavidHurta commented on pkg/updatestatus/nodeinformer_test.go:
```diff
@@ -993,7 +1044,6 @@ func Test_sync_with_node(t *testing.T) {
 					Labels: map[string]string{"node-role.kubernetes.io/some": ""},
 				},
 			},
-			expectedErr: fmt.Errorf("failed to determine which machine config pool the node belongs to: %w", fmt.Errorf("failed to find a matching node selector from 2 machine config pools")),
```
We can now update the name of this test case as an error is not expected at the moment.

hongkailiu replied:
I will keep https://github.com/openshift/cluster-version-operator/pull/1144#discussion_r1938521095 open and close this one as resolved.

DavidHurta commented on pkg/updatestatus/nodeinformer_test.go:
```diff
@@ -1051,3 +1105,140 @@ func Test_sync_with_node(t *testing.T) {
+	mcIndexer := cache.NewIndexer(cache.MetaNamespaceKeyFunc, cache.Indexers{})
+	for _, o := range []metav1.Object{} {
+		if err := mcIndexer.Add(o); err != nil {
+			t.Fatalf("Failed to add o to indexer: %v", err)
```
nit:
```suggestion
			t.Fatalf("Failed to add object to indexer: %v", err)
```

hongkailiu replied:
Done in https://github.com/openshift/cluster-version-operator/compare/791b701aba13b9760aeee99f6ca01e5e2f6a2967..4ff76c0109d0569ce6d4a20042c77d2be29cf5a9

DavidHurta commented on pkg/updatestatus/nodeinformer_test.go:
```diff
@@ -1051,3 +1105,140 @@ func Test_sync_with_node(t *testing.T) {
+	for _, o := range []metav1.Object{&corev1.Node{ObjectMeta: metav1.ObjectMeta{Name: "master-1"}},
+		&corev1.Node{ObjectMeta: metav1.ObjectMeta{Name: "worker-1"}}} {
+		if err := nodeIndexer.Add(o); err != nil {
+			t.Fatalf("Failed to add o to indexer: %v", err)
```
nit:
```suggestion
			t.Fatalf("Failed to add object to indexer: %v", err)
```

hongkailiu replied:
Done in https://github.com/openshift/cluster-version-operator/compare/791b701aba13b9760aeee99f6ca01e5e2f6a2967..4ff76c0109d0569ce6d4a20042c77d2be29cf5a9

DavidHurta commented on pkg/updatestatus/nodeinformer_test.go:
```diff
@@ -1051,3 +1105,140 @@ func Test_sync_with_node(t *testing.T) {
+			mcpToRemove:      "non-exist-mcp",
+			expectedQueueLen: 2,
+		},
+		{
+			name: "reconcile for a new pool",
+			object: &machineconfigv1.MachineConfigPool{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "infra",
+				},
+			},
+			pools:    []*machineconfigv1.MachineConfigPool{getMCP("master"), getMCP("worker")},
+			mcpToAdd: "infra",
+		},
```
Hmm, a reconcile is expected from my understanding but the `expectedQueueLen` is set to zero.

hongkailiu replied:
Good catch. How come I did not notice it. Thanks.

hongkailiu replied:
Done in https://github.com/openshift/cluster-version-operator/compare/791b701aba13b9760aeee99f6ca01e5e2f6a2967..4ff76c0109d0569ce6d4a20042c77d2be29cf5a9

hongkailiu commented on pkg/updatestatus/nodeinformer.go:
```diff
@@ -158,39 +285,28 @@ func makeInsightMsgForNode(nodeInsight *NodeStatusInsight, acquiredAt metav1.Tim
 	}
-	return nil, fmt.Errorf("failed to find a matching node selector from %d machine config pools", len(pools))
+
+	return nil
```
Since it is likely to get more involved to handle the case, I will isolate it to another PR.
I have already remove the item from the PR description.

hongkailiu commented on pkg/updatestatus/nodeinformer.go:
```diff
@@ -125,9 +140,45 @@ func (c *nodeInformerController) sync(ctx context.Context, syncCtx factory.SyncC
+			return c.reconcileAllNodes(syncCtx.Queue())
+		}
+		return nil
 	default:
```
Done in https://github.com/openshift/cluster-version-operator/compare/791b701aba13b9760aeee99f6ca01e5e2f6a2967..4ff76c0109d0569ce6d4a20042c77d2be29cf5a9

hongkailiu commented:
I cannot reproduce the error locally. Strange.

/test lint
/test unit

hongkailiu commented:
/test e2e-hypershift

petr-muller commented:
The PR was a little hard to review with a single commit that does three different things. It would be have been easier if it were three PRs, or at least three commits:

1. Add the ability to requeue all Nodes
2. Add the MC version cache
3. Add the MCP selector cache

petr-muller commented:
I am a bit nervous about the cold start thought. We will receive a random flurry of events, but it does not make much sense processing Nodes until we saw each MCP and each MC at least once, otherwise we end up requeuing a ton of nodes  :thinking:

Maybe we could use a [post-start](https://github.com/openshift/library-go/blob/80620876b7c2dbd5def6da140ad044e6f3de98b5/pkg/controller/factory/factory.go#L182-L186) hook to warmup the caches?

I see you already had something like this and removed it (https://github.com/openshift/cluster-version-operator/pull/1144#issuecomment-2629229145 / https://github.com/openshift/cluster-version-operator/pull/1144#issuecomment-2629433446 ) but I do not entirely understand why you found it useless?

hongkailiu commented on pkg/updatestatus/nodeinformer.go:
```diff
@@ -39,10 +41,23 @@ type nodeInformerController struct {
 
+	// machineConfigPoolCache caches machine config pools
+	machineConfigPoolCache     map[string]machineConfigPoolCacheData
+	machineConfigPoolCacheLock sync.Mutex
```
Done.

hongkailiu commented on pkg/updatestatus/nodeinformer.go:
```diff
@@ -39,10 +41,23 @@ type nodeInformerController struct {
 }
 
+type machineConfigPoolCacheData struct {
+	pool     *machineconfigv1.MachineConfigPool
```
> but we still need to eventually get the given MCP resource from the informer

I thought you wanted me not to do that either in the last round.

(I admit that the controller is smarter with the cache.) Originally I thought that even "the unnecessity to iterate all MCPs" was not a big deal since "the informer also has a cache inside" (thus listing is easy and there are not many MCPs usually) and matching is also not hard either. The code is easier (at least to me) to follow without the cache.

hongkailiu commented on pkg/updatestatus/nodeinformer.go:
```diff
@@ -94,26 +119,11 @@ func (c *nodeInformerController) sync(ctx context.Context, syncCtx factory.SyncC
-			}
+		mcp := c.whichMCP(node)
+		if mcp == nil {
+			return fmt.Errorf("failed to determine which machine config pool the node belongs to")
```
See https://github.com/openshift/cluster-version-operator/pull/1144#discussion_r1955516223

hongkailiu commented:
> I see you already had something like this and removed it ([#1144 (comment)](https://github.com/openshift/cluster-version-operator/pull/1144#issuecomment-2629229145) / [#1144 (comment)](https://github.com/openshift/cluster-version-operator/pull/1144#issuecomment-2629433446) ) but I do not entirely understand why you found it useless?

I could not warm up the cache with the listers from the informer as I do not know when the lister is synced, i.e., they have the MCPs/MCs. My init function was useless because the listers was NOT synced when it is called. Moreover, from the testing, MC/MCP events always came earlier than node events, i.e., "requeuing a ton of nodes" did not happen.

> Maybe we could use a [post-start](https://github.com/openshift/library-go/blob/80620876b7c2dbd5def6da140ad044e6f3de98b5/pkg/controller/factory/factory.go#L182-L186) hook to warmup the caches?

Added it back in the 5th commit with the hook. Let us see if it helps in the test.

If still not helping, `machineConfigInformers.WaitForCacheSync(ctx.Done())` is the next thing to try before this line:

https://github.com/openshift/cluster-version-operator/blob/3af45bc532e17e6ef5abb3eee9c28317714a348d/pkg/updatestatus/start.go#L64

My expectation is that the listers are synced before `nodeInformerController` runs.

hongkailiu commented:
/test e2e-agnostic-usc-devpreview

hongkailiu commented:
Must gather did not work in the last CI run.

/test e2e-agnostic-usc-devpreview

hongkailiu commented:
The ordering looks so nice even without explicitly calling`machineConfigInformers.WaitForCacheSync(ctx.Done())`.

```console
$ curl -s https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/openshift_cluster-version-operator/1144/pull-ci-openshift-cluster-version-operator-main-e2e-agnostic-usc-devpreview/1890400781975162880/artifacts/e2e-agnostic-usc-devpreview/gather-extra/artifacts/pods/openshift-update-status-controller_update-status-controller-7f56cd7468-ffg4q_update-status-controller.log | rg 'Caches are synced|Stored'
I0214 15:31:07.406362       1 shared_informer.go:320] Caches are synced for RequestHeaderAuthRequestController
I0214 15:31:07.406389       1 shared_informer.go:320] Caches are synced for client-ca::kube-system::extension-apiserver-authentication::requestheader-client-ca-file
I0214 15:31:07.406459       1 shared_informer.go:320] Caches are synced for client-ca::kube-system::extension-apiserver-authentication::client-ca-file
I0214 15:31:07.423666       1 base_controller.go:82] Caches are synced for ControlPlaneInformer
I0214 15:31:07.423697       1 base_controller.go:82] Caches are synced for NodeInformer
I0214 15:31:07.423727       1 base_controller.go:82] Caches are synced for UpdateStatusController
I0214 15:31:07.423810       1 nodeinformer.go:209] Stored 2 machineConfigPools in the cache
I0214 15:31:07.424105       1 nodeinformer.go:219] Stored 4 machineConfig versions in the cache
```

Let me try again.

/test e2e-agnostic-usc-devpreview

hongkailiu commented:
install failure in the last CI job.

/test e2e-agnostic-usc-devpreview

hongkailiu commented:
```console
curl -s https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/openshift_cluster-version-operator/1144/pull-ci-openshift-cluster-version-operator-main-e2e-agnostic-usc-devpreview/1890553751542960128/artifacts/e2e-agnostic-usc-devpreview/gather-extra/artifacts/pods/openshift-update-status-controller_update-status-controller-5dc5599f49-6ff4n_update-status-controller.log | rg 'Caches are synced|Stored|Syncing with key|belong'
I0215 01:45:51.099734       1 shared_informer.go:320] Caches are synced for client-ca::kube-system::extension-apiserver-authentication::client-ca-file
I0215 01:45:51.100905       1 shared_informer.go:320] Caches are synced for RequestHeaderAuthRequestController
I0215 01:45:51.100990       1 shared_informer.go:320] Caches are synced for client-ca::kube-system::extension-apiserver-authentication::requestheader-client-ca-file
I0215 01:45:51.117880       1 base_controller.go:82] Caches are synced for NodeInformer
I0215 01:45:51.118257       1 nodeinformer.go:209] Stored 2 machineConfigPools in the cache
I0215 01:45:51.118563       1 nodeinformer.go:219] Stored 4 machineConfig versions in the cache
I0215 01:45:51.118158       1 nodeinformer.go:95] NI :: Syncing with key MachineConfigPool/master
I0215 01:45:51.118719       1 nodeinformer.go:95] NI :: Syncing with key MachineConfigPool/worker
I0215 01:45:51.118781       1 nodeinformer.go:95] NI :: Syncing with key Node/ci-op-4gsljddq-8c750-8g5b5-master-0
I0215 01:45:51.118859       1 nodeinformer.go:122] Node ci-op-4gsljddq-8c750-8g5b5-master-0 belongs to machine config pool master
I0215 01:45:51.118534       1 base_controller.go:82] Caches are synced for UpdateStatusController
I0215 01:45:51.118548       1 base_controller.go:82] Caches are synced for ControlPlaneInformer
I0215 01:45:51.130188       1 nodeinformer.go:95] NI :: Syncing with key Node/ci-op-4gsljddq-8c750-8g5b5-master-1
```

Still good. This time, MCP events came first.

/test e2e-agnostic-usc-devpreview

hongkailiu commented:
```console
curl -s https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/openshift_cluster-version-operator/1144/pull-ci-openshift-cluster-version-operator-main-e2e-agnostic-usc-devpreview/1890602608901492736/artifacts/e2e-agnostic-usc-devpreview/gather-extra/artifacts/pods/openshift-update-status-controller_update-status-controller-76769d4bcc-gdfrq_update-status-controller.log | rg 'Caches are synced|Stored|Syncing with key|belong'
I0215 04:57:44.796093       1 shared_informer.go:320] Caches are synced for client-ca::kube-system::extension-apiserver-authentication::client-ca-file
I0215 04:57:44.796128       1 shared_informer.go:320] Caches are synced for RequestHeaderAuthRequestController
I0215 04:57:44.796132       1 shared_informer.go:320] Caches are synced for client-ca::kube-system::extension-apiserver-authentication::requestheader-client-ca-file
I0215 04:57:44.809388       1 base_controller.go:82] Caches are synced for ControlPlaneInformer
I0215 04:57:44.809416       1 base_controller.go:82] Caches are synced for NodeInformer
I0215 04:57:44.809519       1 nodeinformer.go:209] Stored 2 machineConfigPools in the cache
I0215 04:57:44.809521       1 nodeinformer.go:95] NI :: Syncing with key Node/ci-op-4gsljddq-8c750-2qcn6-master-0
I0215 04:57:44.809519       1 base_controller.go:82] Caches are synced for UpdateStatusController
I0215 04:57:44.809539       1 nodeinformer.go:122] Node ci-op-4gsljddq-8c750-2qcn6-master-0 belongs to machine config pool master
I0215 04:57:44.809561       1 nodeinformer.go:219] Stored 4 machineConfig versions in the cache
I0215 04:57:44.814586       1 nodeinformer.go:95] NI :: Syncing with key Node/ci-op-4gsljddq-8c750-2qcn6-master-1
I0215 04:57:44.814600       1 nodeinformer.go:122] Node ci-op-4gsljddq-8c750-2qcn6-master-1 belongs to machine config pool master
```

Still works: the first node got its matching MCP.

Three consecutive tests worked.

To summarize:
- informer.start(): not blocking
- controller.run() races with posthook.
- controller.run() calls WaitForCacheSync() and posthook needs the synced lister to initialize the caches.

In all three tests so far, the sync had quick enough so that the initializing the caches was either complete (test 1 and 2) or close to complete (test 3) when the node event came in.

I feel this is good enough (the window of cold caches is small) and the MCP/MC event will reconcile all nodes.
We do not have to call WaitForCacheSync explicitly.

Just for my curiosity, let me see the difference if we do it anyway.

hongkailiu commented:
```
curl -s https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/openshift_cluster-version-operator/1144/pull-ci-openshift-cluster-version-operator-main-e2e-agnostic-usc-devpreview/1890735530849603584/artifacts/e2e-agnostic-usc-devpreview/gather-extra/artifacts/pods/openshift-update-status-controller_update-status-controller-6b58f4f84d-gwwjc_update-status-controller.log | rg 'Caches are synced|Stored|Syncing with key|belong'
I0215 13:38:35.934527       1 shared_informer.go:320] Caches are synced for client-ca::kube-system::extension-apiserver-authentication::client-ca-file
I0215 13:38:35.934609       1 shared_informer.go:320] Caches are synced for RequestHeaderAuthRequestController
I0215 13:38:35.934635       1 shared_informer.go:320] Caches are synced for client-ca::kube-system::extension-apiserver-authentication::requestheader-client-ca-file
I0215 13:38:35.954205       1 base_controller.go:82] Caches are synced for NodeInformer
I0215 13:38:35.954048       1 base_controller.go:82] Caches are synced for ControlPlaneInformer
I0215 13:38:35.954334       1 nodeinformer.go:209] Stored 2 machineConfigPools in the cache
I0215 13:38:35.954385       1 nodeinformer.go:219] Stored 4 machineConfig versions in the cache
I0215 13:38:35.954427       1 nodeinformer.go:95] NI :: Syncing with key MachineConfigPool/master
I0215 13:38:35.954477       1 nodeinformer.go:95] NI :: Syncing with key MachineConfigPool/worker
I0215 13:38:35.954493       1 nodeinformer.go:95] NI :: Syncing with key Node/ci-op-19xpkgks-8c750-6xlg8-master-0
I0215 13:38:35.954508       1 nodeinformer.go:122] Node ci-op-19xpkgks-8c750-6xlg8-master-0 belongs to machine config pool master
I0215 13:38:35.954058       1 base_controller.go:82] Caches are synced for UpdateStatusController
I0215 13:38:35.965000       1 nodeinformer.go:95] NI :: Syncing with key Node/ci-op-19xpkgks-8c750-6xlg8-master-1
```

It does what I expect it to do: First event came before caches are initialized.

/test e2e-agnostic-usc-devpreview

Trying again...

hongkailiu commented:
```console
curl -s https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/openshift_cluster-version-operator/1144/pull-ci-openshift-cluster-version-operator-main-e2e-agnostic-usc-devpreview/1890768740337848320/artifacts/e2e-agnostic-usc-devpreview/gather-extra/artifacts/pods/openshift-update-status-controller_update-status-controller-85bcd6d7c6-f4jp7_update-status-controller.log | rg 'Caches are synced|Stored|Syncing with key|belong'
I0215 15:27:40.242112       1 shared_informer.go:320] Caches are synced for client-ca::kube-system::extension-apiserver-authentication::requestheader-client-ca-file
I0215 15:27:40.242132       1 shared_informer.go:320] Caches are synced for RequestHeaderAuthRequestController
I0215 15:27:40.242171       1 shared_informer.go:320] Caches are synced for client-ca::kube-system::extension-apiserver-authentication::client-ca-file
I0215 15:27:40.257544       1 base_controller.go:82] Caches are synced for NodeInformer
I0215 15:27:40.257806       1 nodeinformer.go:209] Stored 2 machineConfigPools in the cache
I0215 15:27:40.257851       1 nodeinformer.go:219] Stored 4 machineConfig versions in the cache
I0215 15:27:40.257670       1 base_controller.go:82] Caches are synced for ControlPlaneInformer
I0215 15:27:40.257675       1 nodeinformer.go:95] NI :: Syncing with key MachineConfigPool/master
I0215 15:27:40.257971       1 nodeinformer.go:95] NI :: Syncing with key MachineConfigPool/worker
I0215 15:27:40.257994       1 nodeinformer.go:95] NI :: Syncing with key Node/ci-op-19xpkgks-8c750-fhcw9-master-0
I0215 15:27:40.258007       1 nodeinformer.go:122] Node ci-op-19xpkgks-8c750-fhcw9-master-0 belongs to machine config pool master
```

Again, like Test 2 above.

/test e2e-agnostic-usc-devpreview

hongkailiu commented:
"gather-extra failed", rerunning ...

/test e2e-agnostic-usc-devpreview

hongkailiu commented:
Something wrong with the cluster. Retrying ...
/test e2e-agnostic-usc-devpreview

hongkailiu commented:
```console
curl -s https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/openshift_cluster-version-operator/1144/pull-ci-openshift-cluster-version-operator-main-e2e-agnostic-usc-devpreview/1890965887544987648/artifacts/e2e-agnostic-usc-devpreview/gather-extra/artifacts/pods/openshift-update-status-controller_update-status-controller-79f559cc9d-qjmlx_update-status-controller.log | rg 'Caches are synced|Stored|Syncing with key|belong'
I0216 04:54:37.613087       1 shared_informer.go:320] Caches are synced for client-ca::kube-system::extension-apiserver-authentication::requestheader-client-ca-file
I0216 04:54:37.613168       1 shared_informer.go:320] Caches are synced for RequestHeaderAuthRequestController
I0216 04:54:37.614167       1 shared_informer.go:320] Caches are synced for client-ca::kube-system::extension-apiserver-authentication::client-ca-file
I0216 04:54:37.625667       1 base_controller.go:82] Caches are synced for NodeInformer
I0216 04:54:37.625830       1 nodeinformer.go:209] Stored 2 machineConfigPools in the cache
I0216 04:54:37.626125       1 nodeinformer.go:219] Stored 4 machineConfig versions in the cache
I0216 04:54:37.626194       1 nodeinformer.go:95] NI :: Syncing with key Node/ci-op-19xpkgks-8c750-8b42p-master-0
I0216 04:54:37.626232       1 nodeinformer.go:122] Node ci-op-19xpkgks-8c750-8b42p-master-0 belongs to machine config pool master
I0216 04:54:37.627826       1 base_controller.go:82] Caches are synced for UpdateStatusController
I0216 04:54:37.633539       1 nodeinformer.go:95] NI :: Syncing with key Node/ci-op-19xpkgks-8c750-8b42p-master-1
I0216 04:54:37.633582       1 nodeinformer.go:122] Node ci-op-19xpkgks-8c750-8b42p-master-1 belongs to machine config pool master
```

Although the CI job failed, we got the logs from the test.
Node event came in first AFTER our caches had been initialized.

We have 3 tests without [7c289e9](https://github.com/openshift/cluster-version-operator/pull/1144/commits/7c289e917740ad42577e64977ce51371579f1212) and 3 with it.

Facts:
- all 6: the informers' caches had been synced before PostStartHook was called.
- all 6: all nodes got a matching MCP.
- 1/3 without it vs 3/3 with it: PostStartHook had finished before the first event came.

Thinking:
- Even without `WaitForCacheSync`, informers' caches were synced. 
- Even with it, it wont ensure that PostStartHook finishes before the first event comes. It is my understanding the code. It did not happen in the testing. It seems it helped to achieve the goal (if the samples are not too few): 1/3 ---> 3/3. So maybe keep the `WaitForCacheSync`?

petr-muller commented:
OK, I proposed `PostStartHook` because I thought they would run before the reconciliation loop but they don't (check out the `Run()` [implementation](https://github.com/openshift/library-go/blob/017e5b6cf27c5cd89620c8a6f3a3faa5f1e8a576/pkg/controller/factory/base_controller.go#L87-L182), it's actually very nice and readable...).

That means `PostStartHook` will really race with our `sync`, which makes it inappropriate solution for the problem. I think we could use something like `sync.Once` (https://pkg.go.dev/sync#Once)  for warmup in `sync()` - first `sync` would call the warmup code at the beginning, and maybe that's enough?

hongkailiu commented:
```console
curl -s https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/openshift_cluster-version-operator/1144/pull-ci-openshift-cluster-version-operator-main-e2e-agnostic-usc-devpreview/1894124084053151744/artifacts/e2e-agnostic-usc-devpreview/gather-extra/artifacts/pods/openshift-update-status-controller_update-status-controller-785c89d6b-dmj26_update-status-controller.log | rg 'Caches are synced|Stored|Syncing with key|belong'
I0224 21:53:16.289576       1 shared_informer.go:320] Caches are synced for client-ca::kube-system::extension-apiserver-authentication::requestheader-client-ca-file
I0224 21:53:16.289608       1 shared_informer.go:320] Caches are synced for client-ca::kube-system::extension-apiserver-authentication::client-ca-file
I0224 21:53:16.289585       1 shared_informer.go:320] Caches are synced for RequestHeaderAuthRequestController
I0224 21:53:16.306588       1 base_controller.go:82] Caches are synced for ControlPlaneInformer
I0224 21:53:16.306599       1 base_controller.go:82] Caches are synced for NodeInformer
I0224 21:53:16.306626       1 base_controller.go:82] Caches are synced for UpdateStatusController
I0224 21:53:16.306836       1 nodeinformer.go:221] Stored 2 machineConfigPools in the cache
I0224 21:53:16.306874       1 nodeinformer.go:231] Stored 2 machineConfig versions in the cache
I0224 21:53:16.306880       1 nodeinformer.go:101] NI :: Syncing with key MachineConfigPool/master
I0224 21:53:16.306900       1 nodeinformer.go:101] NI :: Syncing with key MachineConfigPool/worker
I0224 21:53:16.306916       1 nodeinformer.go:101] NI :: Syncing with key Node/ci-op-yc00jphy-8c750-78krc-master-0
I0224 21:53:16.306930       1 nodeinformer.go:128] Node ci-op-yc00jphy-8c750-78krc-master-0 belongs to machine config pool master
...
```

The above log is from [this job](https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_cluster-version-operator/1144/pull-ci-openshift-cluster-version-operator-main-e2e-agnostic-usc-devpreview/1894124084053151744). It (`once.Do(c.initializeCachesNoErrors)`) seems working nicely: called ONCE and ONLY once.

hongkailiu commented:
/retest-required

hongkailiu commented:
/retest-required

petr-muller commented on pkg/updatestatus/nodeinformer.go:
```diff
@@ -155,6 +165,52 @@ func (c *nodeInformerController) sync(ctx context.Context, syncCtx factory.SyncC
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	v, ok := c.cache[mc.Name]
+	if openshiftVersion, exist := mc.Annotations[mco.ReleaseImageVersionAnnotationKey]; exist && openshiftVersion != "" {
```
nit: I think we could benefit from some naming symmetry here, like:

```
cachedVersion, cached := c.cache[name]
mcVersion, annotated := mc.Annotations[mco.ReleaseImageVersionAnnotationKey]
if annotated && mcVersion != "" {
  if !cached || cachedVersion != mcVersion {
  ...
  }
} else if cached {
}
````

hongkailiu replied:
Done.
https://github.com/openshift/cluster-version-operator/pull/1164/commits/a81ec69e31f504ce10a44650cfa5c4d4260fd2cd#diff-c5a41cb83200e13544e0509a5e77c180e16a1d70d7411a79cac2bc107e89950fR178

petr-muller commented on pkg/updatestatus/nodeinformer.go:
```diff
@@ -343,16 +399,16 @@ func toPointer(d time.Duration) *metav1.Duration {
 }
 
-func assessNode(node *corev1.Node, mcp *machineconfigv1.MachineConfigPool, machineConfigVersions map[string]string, mostRecentVersionInCVHistory string, now metav1.Time) *updatestatus.NodeStatusInsight {
+func assessNode(node *corev1.Node, mcp *machineconfigv1.MachineConfigPool, machineConfigVersionMatcher func(string) (string, bool), mostRecentVersionInCVHistory string, now metav1.Time) *updatestatus.NodeStatusInsight {
```
nit: `...matcher` does not tell much about the what is the input and output, how about `versionForMachineConfig` or `machineConfigToVersion`

hongkailiu replied:
I chose `machineConfigToVersion`.

https://github.com/openshift/cluster-version-operator/pull/1164/commits/a81ec69e31f504ce10a44650cfa5c4d4260fd2cd#diff-c5a41cb83200e13544e0509a5e77c180e16a1d70d7411a79cac2bc107e89950fR523

petr-muller commented on pkg/updatestatus/nodeinformer.go:
```diff
@@ -155,6 +165,52 @@ func (c *nodeInformerController) sync(ctx context.Context, syncCtx factory.SyncC
+	return false
+}
+
+func (c *machineConfigVersionCache) match(config string) (string, bool) {
```
nit: consider `versionFor(name string)`

tells more directly what is the output, and param `name` is consistent with `forget`

hongkailiu replied:
Done.

https://github.com/openshift/cluster-version-operator/pull/1164/commits/a81ec69e31f504ce10a44650cfa5c4d4260fd2cd#diff-c5a41cb83200e13544e0509a5e77c180e16a1d70d7411a79cac2bc107e89950fR203

petr-muller commented on pkg/updatestatus/nodeinformer_test.go:
```diff
@@ -149,6 +151,23 @@ func getMCP(name string) *machineconfigv1.MachineConfigPool {
 	}
 }
 
+func (c *nodeInformerController) initializeCaches() error {
```
placing `nodeInformerController` method in `_test.go` file does not look right

hongkailiu replied:
I think It is indeed a bug. I made a separate commit for its fix.

https://github.com/openshift/cluster-version-operator/pull/1164/commits/998d96cb2a0bb8f80fcf40df15673953fd1180c3

petr-muller commented on pkg/updatestatus/nodeinformer.go:
```diff
@@ -40,6 +40,9 @@ type nodeInformerController struct {
 	sendInsight sendInsightFn
 
+	// machineConfigPoolSelectorCache caches the label selectors converted from the node selectors of the machine config pools by their names.
+	machineConfigPoolSelectorCache machineConfigPoolSelectorCache
```
consider calling the member just `mcpSelectors` (type name is good though)

hongkailiu replied:
Done.

https://github.com/openshift/cluster-version-operator/pull/1164/commits/a81ec69e31f504ce10a44650cfa5c4d4260fd2cd#diff-c5a41cb83200e13544e0509a5e77c180e16a1d70d7411a79cac2bc107e89950fR43

petr-muller commented on pkg/updatestatus/nodeinformer.go:
```diff
@@ -152,7 +161,23 @@ func (c *nodeInformerController) sync(ctx context.Context, syncCtx factory.SyncC
+		if err != nil && !kerrors.IsNotFound(err) {
+			return err
+		}
+		if kerrors.IsNotFound(err) {
+			// The pool was deleted
+			if changed := c.machineConfigPoolSelectorCache.forget(name); changed {
+				klog.V(2).Infof("Reconciling all nodes as machine config pool %q is deleted", name)
+				return c.reconcileAllNodes(syncCtx.Queue())
+			}
+			return nil
+		}
+		if changed := c.machineConfigPoolSelectorCache.ingest(machineConfigPool); changed {
+			klog.V(2).Infof("Reconciling all nodes as machine config pool %q is refreshed", name)
+			return c.reconcileAllNodes(syncCtx.Queue())
+		}
```
consider similar structural changes I suggested for MC one

hongkailiu replied:
Done.

https://github.com/openshift/cluster-version-operator/pull/1164/commits/a81ec69e31f504ce10a44650cfa5c4d4260fd2cd#diff-c5a41cb83200e13544e0509a5e77c180e16a1d70d7411a79cac2bc107e89950fR295

petr-muller commented on pkg/updatestatus/nodeinformer.go:
```diff
@@ -211,6 +236,73 @@ func (c *machineConfigVersionCache) match(config string) (string, bool) {
 	return v, ok
 }
 
+type machineConfigPoolSelectorCache struct {
+	cache map[string]labels.Selector
+	lock  sync.Mutex
+}
```
consider using `sync.Map` like I suggested for MC one

hongkailiu replied:
Done.

https://github.com/openshift/cluster-version-operator/pull/1164/commits/a81ec69e31f504ce10a44650cfa5c4d4260fd2cd#diff-c5a41cb83200e13544e0509a5e77c180e16a1d70d7411a79cac2bc107e89950fR212

petr-muller commented on pkg/updatestatus/nodeinformer.go:
```diff
@@ -303,6 +307,12 @@ func (c *machineConfigPoolSelectorCache) forget(mcpName string) bool {
 	return false
 }
 
+func (c *machineConfigPoolSelectorCache) len() int {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	return len(c.cache)
+}
```
Feels like overkill to have a locked method just to log `...N config pools` - I think a simple "does not belong to any MachineConfigPool" log is just as useful :shrug:

petr-muller commented on pkg/updatestatus/nodeinformer.go:
```diff
@@ -194,6 +202,37 @@ func (c *nodeInformerController) sync(ctx context.Context, syncCtx factory.SyncC
+	klog.V(2).Infof("Stored %d machineConfig versions in the cache", c.machineConfigVersionCache.len())
+
+	return kutilerrors.NewAggregate(errs)
+}
```
I'd prefer to not couple the two into a single method: they are similar but are used in a different way at different places - consider having separate dedicated methods (which saves us from `NewAggregate` so the two methods would even be simpler, and have two `Once` instances.

hongkailiu replied:
Two init methods now but still they are called in the same Once instance.

https://github.com/openshift/cluster-version-operator/pull/1164/commits/a81ec69e31f504ce10a44650cfa5c4d4260fd2cd#diff-c5a41cb83200e13544e0509a5e77c180e16a1d70d7411a79cac2bc107e89950fR99

petr-muller commented on pkg/updatestatus/nodeinformer.go:
```diff
@@ -194,6 +202,37 @@ func (c *nodeInformerController) sync(ctx context.Context, syncCtx factory.SyncC
+			c.machineConfigPoolSelectorCache.ingest(pool)
+		}
+	}
+	klog.V(2).Infof("Stored %d machineConfigPools in the cache", c.machineConfigPoolSelectorCache.len())
```
consider not having a `len()` (that needs to lock) and just count the ingestions?

hongkailiu replied:
Done.

https://github.com/openshift/cluster-version-operator/pull/1164/commits/a81ec69e31f504ce10a44650cfa5c4d4260fd2cd#diff-c5a41cb83200e13544e0509a5e77c180e16a1d70d7411a79cac2bc107e89950fR155

petr-muller commented on pkg/updatestatus/nodeinformer.go:
```diff
@@ -194,6 +202,37 @@ func (c *nodeInformerController) sync(ctx context.Context, syncCtx factory.SyncC
+	if err := c.initializeCaches(); err != nil {
+		klog.Errorf("Failed to initialize caches: %v", err)
+	}
+}
```
I wondered whether we'd need to abort the controller :thinking: 

I _think_ it is safe to continue the controller despite these failures because it just means we started cold and will eventually load all items from the cache?

hongkailiu replied:
Done.

https://github.com/openshift/cluster-version-operator/pull/1164/commits/a81ec69e31f504ce10a44650cfa5c4d4260fd2cd#diff-c5a41cb83200e13544e0509a5e77c180e16a1d70d7411a79cac2bc107e89950fR99

I am not 100% sure if I got your suggestion correctly and keep this unsolved so it is more visible to you to get another look.

petr-muller commented on pkg/updatestatus/nodeinformer.go:
```diff
@@ -132,15 +125,32 @@ func (c *nodeInformerController) sync(ctx context.Context, syncCtx factory.SyncC
+			if changed := c.machineConfigVersionCache.forget(name); changed {
+
+				klog.V(2).Infof("Reconciling all nodes as machine config %q is deleted", name)
+				return c.reconcileAllNodes(syncCtx.Queue())
```
I think we are losing some robustness by directly returning an error here. Consider the following;

1. MC is updated and carries a new version - cache returns `changed` and we get into one of these branches where we do `reconcileAll`
2. The `List` inside `reconcileAll` fails so we propagate the err outside, and the controller library will requeue the MC queueKey that triggered us
3. New `sync` is called for this MC - but this time the cache already has the version, so `changed==false` and we finish the sync without calling `reconcileAll` -> and end up with potentially outdated insights :fearful: 

I'm not sure if it is an ugly hack or not, but I wonder if `reconcileAllNodes` could be a special synthetic queue key that we'd just queue here, and it would get its own dedicated `sync()` call. That synthetic call would then be handled simply by what `reconcileAllNodes` is doing, and on a List error it would just return it, which would cause a requeue of the `reconcileAllNodes` key... WDYT?

hongkailiu replied:
Done with [this comment](https://github.com/openshift/cluster-version-operator/pull/1164/commits/4bebc6254afd513dcbac64f8fb690649b7061a47).

petr-muller reviewed (commented):
Many small bits of feedback inline. I think the only substantial change is the lack of robustness when `reconcileAllNodes` fails. The rest is mostly naming and code structure proposals - address the ones you agree with.

In line with our retro agreement, I am approving the PR so you can merge it and address the feedback in a followup.

/hold
/lgtm
/approve
/label no-qe

petr-muller commented:
/lgtm
/approve

hongkailiu commented:
I will address the comments by separate PRs.

/hold cancel

hongkailiu commented:
/retest-required

hongkailiu commented:
/retest-required

hongkailiu commented:
/retest-required

hongkailiu commented:
/test e2e-agnostic-ovn

hongkailiu commented on pkg/updatestatus/nodeinformer.go:
```diff
@@ -39,6 +40,10 @@ type nodeInformerController struct {
 
+	// machineConfigVersionCache caches machine config versions
+	// The cache stores the name of MC as the key and the release image version as its value which is retrieved from the annotation of the MC.
+	machineConfigVersionCache machineConfigVersionCache
```
I chose [the other suggestion](https://github.com/openshift/cluster-version-operator/pull/1144#discussion_r1969810868).

hongkailiu commented on pkg/updatestatus/nodeinformer.go:
```diff
@@ -392,8 +418,16 @@ func nodeInformerControllerQueueKeys(object runtime.Object) []string {
 	}
 }
+
+func kindAndNameToQueueKey(kind, name string) string {
```
Done. https://github.com/openshift/cluster-version-operator/pull/1164/commits/a81ec69e31f504ce10a44650cfa5c4d4260fd2cd#diff-c5a41cb83200e13544e0509a5e77c180e16a1d70d7411a79cac2bc107e89950fR615

hongkailiu commented on pkg/updatestatus/nodeinformer.go:
```diff
@@ -39,6 +40,10 @@ type nodeInformerController struct {
 
+	// machineConfigVersionCache caches machine config versions
+	// The cache stores the name of MC as the key and the release image version as its value which is retrieved from the annotation of the MC.
+	machineConfigVersionCache machineConfigVersionCache
```
Done.
https://github.com/openshift/cluster-version-operator/pull/1164/commits/a81ec69e31f504ce10a44650cfa5c4d4260fd2cd#diff-c5a41cb83200e13544e0509a5e77c180e16a1d70d7411a79cac2bc107e89950fR48

hongkailiu commented on pkg/updatestatus/nodeinformer.go:
```diff
@@ -132,15 +125,32 @@ func (c *nodeInformerController) sync(ctx context.Context, syncCtx factory.SyncC
 		}
 	case machineConfigKindName:
-		return c.reconcileAllNodes(syncCtx.Queue())
+		machineConfig, err := c.machineConfigs.Get(name)
+		if err != nil && !kerrors.IsNotFound(err) {
+			return err
+		}
+		if kerrors.IsNotFound(err) {
+			// The machine config was deleted
+			if changed := c.machineConfigVersionCache.forget(name); changed {
+
+				klog.V(2).Infof("Reconciling all nodes as machine config %q is deleted", name)
+				return c.reconcileAllNodes(syncCtx.Queue())
+			}
+			return nil
+		}
+		if changed := c.machineConfigVersionCache.ingest(machineConfig); changed {
+			klog.V(2).Infof("Reconciling all nodes as machine config %q is refreshed", name)
+			return c.reconcileAllNodes(syncCtx.Queue())
+		}
+		return nil
```
Done with [syncMachineConfig](https://github.com/openshift/cluster-version-operator/pull/1164/commits/a81ec69e31f504ce10a44650cfa5c4d4260fd2cd#diff-c5a41cb83200e13544e0509a5e77c180e16a1d70d7411a79cac2bc107e89950fR132), [syncMachineConfigPool](https://github.com/openshift/cluster-version-operator/pull/1164/commits/a81ec69e31f504ce10a44650cfa5c4d4260fd2cd#diff-c5a41cb83200e13544e0509a5e77c180e16a1d70d7411a79cac2bc107e89950fR134), and [syncNode](https://github.com/openshift/cluster-version-operator/pull/1164/commits/a81ec69e31f504ce10a44650cfa5c4d4260fd2cd#diff-c5a41cb83200e13544e0509a5e77c180e16a1d70d7411a79cac2bc107e89950fR124).

hongkailiu commented on pkg/updatestatus/nodeinformer.go:
```diff
@@ -132,15 +125,32 @@ func (c *nodeInformerController) sync(ctx context.Context, syncCtx factory.SyncC
+			return nil
+		}
+		if changed := c.machineConfigVersionCache.ingest(machineConfig); changed {
+			klog.V(2).Infof("Reconciling all nodes as machine config %q is refreshed", name)
```
Done.
https://github.com/openshift/cluster-version-operator/pull/1164/commits/a81ec69e31f504ce10a44650cfa5c4d4260fd2cd#diff-c5a41cb83200e13544e0509a5e77c180e16a1d70d7411a79cac2bc107e89950fR275

hongkailiu commented on pkg/updatestatus/nodeinformer.go:
```diff
@@ -132,15 +125,32 @@ func (c *nodeInformerController) sync(ctx context.Context, syncCtx factory.SyncC
+		if err != nil && !kerrors.IsNotFound(err) {
+			return err
+		}
+		if kerrors.IsNotFound(err) {
+			// The machine config was deleted
+			if changed := c.machineConfigVersionCache.forget(name); changed {
+
+				klog.V(2).Infof("Reconciling all nodes as machine config %q is deleted", name)
+				return c.reconcileAllNodes(syncCtx.Queue())
+			}
+			return nil
+		}
+		if changed := c.machineConfigVersionCache.ingest(machineConfig); changed {
+			klog.V(2).Infof("Reconciling all nodes as machine config %q is refreshed", name)
+			return c.reconcileAllNodes(syncCtx.Queue())
+		}
```
Done.

hongkailiu commented on pkg/updatestatus/nodeinformer.go:
```diff
@@ -155,6 +165,52 @@ func (c *nodeInformerController) sync(ctx context.Context, syncCtx factory.SyncC
 	return nil
 }
 
+type machineConfigVersionCache struct {
+	cache map[string]string
+	lock  sync.Mutex
+}
```
Done.
https://github.com/openshift/cluster-version-operator/pull/1164/commits/a81ec69e31f504ce10a44650cfa5c4d4260fd2cd#diff-c5a41cb83200e13544e0509a5e77c180e16a1d70d7411a79cac2bc107e89950fR173

The API is awesome.

hongkailiu commented on pkg/updatestatus/nodeinformer.go:
```diff
@@ -143,6 +155,17 @@ func (c *nodeInformerController) sync(ctx context.Context, syncCtx factory.SyncC
 }
 
+func (c *nodeInformerController) reconcileAllNodes(queue workqueue.TypedRateLimitingInterface[any]) error {
+	nodes, err := c.nodes.List(labels.Everything())
+	if err != nil {
+		return err
+	}
```
Not returning error any more.
I feel https://github.com/openshift/cluster-version-operator/pull/1144#discussion_r1969918640 makes more sense.

https://github.com/openshift/cluster-version-operator/pull/1164/commits/a81ec69e31f504ce10a44650cfa5c4d4260fd2cd#diff-c5a41cb83200e13544e0509a5e77c180e16a1d70d7411a79cac2bc107e89950fR99
